boefjes/boefjes/plugins/kat_manual/yml/normalize.py

```diff
@@ -0,0 +1,210 @@
+    "SoftwareInstance": {"type": SoftwareInstance, "distinctive_fields": SoftwareInstance._natural_key_attrs},
+    # Web
+    "Website": {"type": Website, "distinctive_fields": Website._natural_key_attrs},
+    # ?baseclass :: referenced at URL, RESTAPI ... :: need to initialize with proper class
+    "WebURL": {"type": WebURL, "distinctive_fields": ["scheme", "port", "path"]},
+    "HostnameHTTPURL": {"type": HostnameHTTPURL, "distinctive_fields": HostnameHTTPURL._natural_key_attrs},
+    "IPAddressHTTPURL": {"type": IPAddressHTTPURL, "distinctive_fields": IPAddressHTTPURL._natural_key_attrs},
+    "HTTPResource": {"type": HTTPResource, "distinctive_fields": HTTPResource._natural_key_attrs},
+    "HTTPHeader": {"type": HTTPHeader, "distinctive_fields": HTTPHeader._natural_key_attrs},
+    "URL": {"type": URL, "distinctive_fields": URL._natural_key_attrs},
+    "HTTPHeaderURL": {"type": HTTPHeaderURL, "distinctive_fields": HTTPHeaderURL._natural_key_attrs},
+    "HTTPHeaderHostname": {"type": HTTPHeaderHostname, "distinctive_fields": HTTPHeaderHostname._natural_key_attrs},
+    "ImageMetadata": {"type": ImageMetadata, "distinctive_fields": ImageMetadata._natural_key_attrs},
+    "RESTAPI": {"type": RESTAPI, "distinctive_fields": RESTAPI._natural_key_attrs},
+    "APIDesignRule": {"type": APIDesignRule, "distinctive_fields": APIDesignRule._natural_key_attrs},
+    "APIDesignRuleResult": {"type": APIDesignRuleResult, "distinctive_fields": APIDesignRuleResult._natural_key_attrs},
+    "SecurityTXT": {"type": SecurityTXT, "distinctive_fields": SecurityTXT._natural_key_attrs},
+}
+
+logger = logging.getLogger(__name__)
+
+
+def run(input_ooi: dict, raw: bytes) -> Iterable[NormalizerOutput]:
+    reference_cache = {"Network": {"internet": Network(name="internet")}}
+
+    yield from process_yml(raw, reference_cache)
+
+
+def process_yml(yml_raw_data: bytes, reference_cache: dict) -> Iterable[NormalizerOutput]:
+    yml_data = io.StringIO(yml_raw_data.decode("UTF-8"))
+    oois_from_yaml = yaml.safe_load(yml_data)
+    oois = []
+    for ooi_number, ooi_dict in enumerate(oois_from_yaml, start=1):
+        try:
+            create_oois(ooi_dict, reference_cache, oois)
+        except ValidationError as err:
+            logger.exception("Validation failed for indexed object at %s", ooi_number)
+            logger.exception(f"with error: {str(err)}")
+    return oois
+
+def create_oois(ooi_dict:dict, reference_cache:dict, oois_list:list):
+    # constants
+    skip_properties = ("object_type", "scan_profile", "primary_key", "user_id")
+    # check for main ooi
+    ooi_type = OOI_TYPES[ooi_dict["ooi_type"]]["type"]
+    # Special Cases
+    if hasattr(ooi_type, 'type_from_raw'): ooi_type = ooi_type.type_from_raw(ooi_dict)
+    # check for cache
+    dins_fields = OOI_TYPES[ooi_type.__name__].get('distinctive_fields')
+    if dins_fields:
+        field_name = get_cache_name(ooi_dict, OOI_TYPES[ooi_type.__name__].get('distinctive_fields'))
```

Comment:
```suggestion
    dins_fields = ooi_type._natural_key_attrs
    if dins_fields:
        field_name = get_cache_name(ooi_dict, dins_fields)
```
This would remove the need for the distinctive_fields in the list OOI_TYPES, and makes sure we use the same fields we also use for the PK in the model, without adding maintenance work for  newer ooi types.
Maybe, the whole OOI_TYPES list can be refactored out by using the available OOI type lists.

Reply:
Yes, repetition is partially fixed.